bundles/org.openhab.binding.e3dc/src/main/java/org/openhab/binding/e3dc/internal/modbus/ModbusCallback.java

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.e3dc.internal.modbus;
+
+import static org.openhab.binding.e3dc.internal.modbus.E3DCModbusConstans.*;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.Optional;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.e3dc.internal.dto.EmergencyBlock;
+import org.openhab.binding.e3dc.internal.dto.InfoBlock;
+import org.openhab.binding.e3dc.internal.dto.PowerBlock;
+import org.openhab.binding.e3dc.internal.dto.StringBlock;
+import org.openhab.binding.e3dc.internal.dto.WallboxArray;
+import org.openhab.binding.e3dc.internal.modbus.Data.DataType;
+import org.openhab.io.transport.modbus.AsyncModbusFailure;
+import org.openhab.io.transport.modbus.AsyncModbusReadResult;
+import org.openhab.io.transport.modbus.ModbusFailureCallback;
+import org.openhab.io.transport.modbus.ModbusReadCallback;
+import org.openhab.io.transport.modbus.ModbusReadRequestBlueprint;
+import org.openhab.io.transport.modbus.ModbusRegister;
+import org.openhab.io.transport.modbus.ModbusRegisterArray;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link InfoBlockCallback} class receives callbacks from modbus poller
+ *
+ * @author Bernd Weymann - Initial contribution
+ */
+@NonNullByDefault
+public class ModbusCallback extends ModbusDataProvider
+        implements ModbusReadCallback, ModbusFailureCallback<ModbusReadRequestBlueprint> {
+    private final Logger logger = LoggerFactory.getLogger(ModbusCallback.class);
+    private DataType callbackType;
+    private byte[] bArray;
+    private int size;
+    private int counter = 0;
+    private long maxDuration = Long.MIN_VALUE;
+    private long minDuration = Long.MAX_VALUE;
+    private long avgDuration = 0;
+
+    public ModbusCallback(DataType type) {
+        callbackType = type;
+        if (type.equals(DataType.INFO)) {
+            size = INFO_REG_SIZE * 2;
+            bArray = new byte[size];
+        } else {
+            size = (REGISTER_LENGTH - INFO_REG_SIZE) * 2;
+            bArray = new byte[size];
+        }
+    }
+
+    @Override
+    public void handle(AsyncModbusReadResult result) {
+        byte[] newArray = new byte[size];
+        long startTime = System.currentTimeMillis();
+        Optional<ModbusRegisterArray> opt = result.getRegisters();
+        ModbusRegisterArray registers = opt.get();
```

Comment:
I am not sure when this callback is registered and what type of requests but `opt.ifPresent(...)` might be preferred to be on the safe side? (see e.g. sunspec: https://github.com/openhab/openhab-addons/blob/2.5.x/bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java#L337)

Non-register payload (e.g. coil data) will now crash with `Optional.get` since the Optional will be empty.

Reply:
Reworked due to change towards Modbus binding